docs/book/common_code/machine_controller.md

```diff
@@ -0,0 +1,265 @@
+	// MachineSet controller will be taking these conditions as a signal to decide if
+	// machine is healthy or needs to be replaced.
+	// Refer: https://kubernetes.io/docs/concepts/architecture/nodes/#condition
+	// +optional
+	Conditions []corev1.NodeCondition `json:"conditions,omitempty"`
+}
+```
+{% endmethod %}
+
+{% method %}
+## Machine Actuator Interface
+
+All methods should be idempotent. Each time the Machine controller attempts
+to reconcile the state it will call one or more of the following actuator
+methods.
+
+`Create()` will only be called when `Exists()` returns false.
+
+`Update()` will only be called when `Exists()` returns true.
+
+`Delete()` will only be called when the `Machine` is in the process of being
+deleted.
+
+The definition of `Exist()` is determined by the provider.
+
+**TODO**: Provide more guidance on `Exists()`.
+
+{% sample lang="go" %}
+```go
+// Actuator controls machines on a specific infrastructure. All
+// methods should be idempotent unless otherwise specified.
+type Actuator interface {
+	// Create the machine.
+	Create(*clusterv1.Cluster, *clusterv1.Machine) error
+	// Delete the machine. If no error is returned, it is assumed that all dependent resources have been cleaned up.
+	Delete(*clusterv1.Cluster, *clusterv1.Machine) error
+	// Update the machine to the provided definition.
+	Update(*clusterv1.Cluster, *clusterv1.Machine) error
+	// Checks if the machine currently exists.
+	Exists(*clusterv1.Cluster, *clusterv1.Machine) (bool, error)
+}
+```
+{% endmethod %}
+
+## Machine Controller Semantics
+
+{% panel style="info", title="Logic sequence" %}
+We need a diagram tracing the logic from resource creation though updates
+and finally deletion.
+{% endpanel %}
+
+0. Determine the `Cluster` associated with the `Machine`.
+- If the `Machine` hasn't been deleted and doesn't have a finalizer, add one.
+- If the `Machine` is being deleted, and there is no finalizer, we're done
+  - Check if the `Machine` is allowed to be deleted. [^1]
+  - Call the provider specific actuators `Delete()` method.
+    - If the `Delete()` method returns true, remove the finalizer.
+- Check if the `Machine` exists by calling the provider specific `Exists()`
+method.
+  - If it does, call the `Update()` method.
+  - If the `Update()` fails and returns a retryable error:
+    - Retry the `Update()` after N seconds.
+- If the machine does not exist, attempt to create machine by calling
+  actuator `Create()` method.
+
+{% panel style="warning", title="Machines depend on Clusters" %}
+The Machine actuator methods expect both a `Cluster` and a `Machine` to be
+passed in. While there is not a strong link between `Cluster`s and `Machine`s,
+the machine controller will determine which cluster to pass by looking for a 
+`Cluster` in the same namespace as the `Machine`
+
+There are two consequences of this:
+ - There must be at exactly one `Cluster` per namespace.
```

Comment:
I'd suggest rewording this to say "There must no be more than one `Cluster` per namespace". The current wording seems to suggest that namespaces without a `cluster` would also be problematic.